### ISSUE TITLE
Allow users to pass label selectors to determine if a resource is eligible for deletion

### ISSUE THREAD

**Issue by alan-ghelardi** [tektoncd/results#286]
### Feature request

Currently, the Watcher deletes a resource when the `completed_run_grace_period` flag is set and the time elapsed since the resource's completion matches the specified value. However, sometimes, users may set  up other tools to watch resources such as PipelineRuns and TaskRuns to perform certain actions. In those cases, it's  useful to wait until those tools complete their job before deleting the resources.

The Watcher could receive a new flag (say, `wait_for_labels`) to set a  list of label selectors to be matched against PipelineRuns and TaskRuns to determine whether the resource in question is eligible to be deleted. External tools that need to perform actions during the lifecycle of those resources must mark them with custom labels to signal to the Watcher that they have completed their job and the resources are eligible for deletion.

### Use case

This feature provides a flexible and decoupled way to external tools to communicate with the Results Watcher in order to synchronize their jobs. For instance, in our company, we've got a controller that stores completed runs in  our data lake. In order  to synchronize the actions performed by this controller and the Results watcher,  our controller would mark the stored runs with a custom label. The Results Watcher would be configured to wait until this label is present to delete the resources. The same principle is applicable to other integrations that can be developed eventually.

Besides of that, this feature would give people a more fine-grained control over deleted runs. If for some reason, users want to preserve certain PipelineRuns in the Kubernetes cluster even after archiving them in the Results API server, they just need to specify label selectors that match those runs.

**Comment by adambkaplan** [comment]
@alan-ghelardi would a [finalizer](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) work better for your situation? That lets controllers do a few things:

1. Indicate an object should not be deleted because some other work is being done.
2. Allow the controller to do any clean up work when an object needs to be deleted.

That said, I can understand why a finalizer would not be suitable as it is very easy for finalizers to be "orphaned" (ex: controller is stopped or upgraded to no longer use a finalizer). This can block cleanup of cluster resources.

**Comment by alan-ghelardi** [comment]
@adambkaplan  usually I prefer to let the controller that owns the resource to deal with finalizers and the approach based on selectors looks like less intrusive and causes less friction in scenarios in which an incident occurs. For instance, if something causes tens of thousands of objects to get accumulated in the cluster and an administrator needs to delete them manually, they need to remove the finalizers to force the objects to be deleted.
In the past, the controller that I mentioned added a finalizer to PipelineRuns, but I decided to remove it because we ran into the exact issue I described. An argument in favor of adding the aforementioned feature to Results is making the process of choosing the resources to be deleted more flexible: one may decide to preserve a certain group of runs in the cluster based  on selectors, for instance. But I'm totally open to give up if you consider that this feature is irrelevant.
